File: tests/js/keymap.js
.import "../../ui/js/Keymap.js" as Keymap

function run(check) {
    var none = Qt.NoModifier, ctrl = Qt.ControlModifier, shift = Qt.ShiftModifier
    var alt = Qt.AltModifier, meta = Qt.MetaModifier
    function key(preset, name, text, mods, expected, context, frontend) {
        check(preset + " " + (context || "listing") + " " + (frontend || "gui") + " " + name + " " + text + " " + mods,
              Keymap.lookupFor(preset, Qt["Key_" + name] || 0, text, mods, context, frontend), expected)
    }
    for (var i = 0; i < Keymap.PRESETS.length; i++) {
        var preset = Keymap.PRESETS[i]
        key(preset, "Down", "", none, "cursorDown")
        key(preset, "Up", "", none, "cursorUp")
        key(preset, "Space", " ", none, "preview")
        key(preset, "P", "p", alt, "togglePreview")
        key(preset, "Space", " ", ctrl, "loadPreview")
        key(preset, "PageDown", "", ctrl, "tabNext")
        key(preset, "PageUp", "", ctrl, "tabPrevious")
        key(preset, "Tab", "", ctrl, "focusPreview")
        key(preset, "Tab", "", none, "focusNext")
        key(preset, "W", "", ctrl, "tabClose")
        key(preset, "S", "s", none, "sortNext")
        key(preset, "S", "S", shift, "sortReverse")
        key(preset, "Slash", "/", none, "filter")
        key(preset, "Question", "?", shift, "keymapSheet")
        key(preset, "Insert", "", ctrl, "copy", "listing", "tui")
        key(preset, "Insert", "", shift, "paste", "listing", "tui")
        key(preset, "Insert", "", ctrl, "", "listing", "gui")
        key(preset, "1", "1", none, "tab1", "listing", "tui")
        key(preset, "9", "9", none, "tab9", "listing", "tui")
        key(preset, "1", "1", none, "", "listing", "gui")
        key(preset, "N", "", preset === "mac" ? meta : ctrl, "windowNew")
        key(preset, "N", "", preset === "mac" ? meta : ctrl, "", "listing", "tui")
        key(preset, "1", "", preset === "windows" ? ctrl | shift : ctrl, "viewList")
        key(preset, "2", "", preset === "windows" ? ctrl | shift : ctrl, "viewColumns")
        key(preset, "3", "", preset === "windows" ? ctrl | shift : ctrl, "viewGrid")
        key(preset, "Comma", "", ctrl, "settings")
        for (var f = 0; f < 2; f++) {
            var frontend = ["gui", "tui"][f]
            var menuContexts = ["listing", "rail", "menu", "panel", "preview", "pdf", "media", "editor"]
            for (var m = 0; m < menuContexts.length; m++) {
                var menuContext = menuContexts[m], menuAction = m < 2 ? "menu" : ""
                key(preset, "Menu", "", none, menuAction, menuContext, frontend)
                key(preset, "F10", "", shift, menuAction, menuContext, frontend)
            }
        }
        key(preset, "N", "", ctrl | shift, "newFolder")
        key(preset, "Q", "q", alt, "")
        key(preset, "J", "j", meta, "")
        key(preset, "D", "D", ctrl | shift, "")
        for (var c = 0; c < 5; c++) {
            var context = ["menu", "panel", "preview", "pdf", "media"][c]
            key(preset, "Return", "", none, "open", context)
            key(preset, "Enter", "", none, "open", context)
            key(preset, "Space", " ", none, "preview", context)
            key(preset, "Escape", "", none, "escape", context)
            key(preset, "D", "d", none, "", context)
            key(preset, "X", "", ctrl, "", context)
            key(preset, "N", "", ctrl, "", context)
        }
        key(preset, "Right", "", none, "menuRight", "menu")
        key(preset, "Left", "", none, "parent", "menu")
        key(preset, "J", "j", none, "cursorDown", "menu")
        key(preset, "K", "k", none, "cursorUp", "menu")
        key(preset, "Minus", "-", none, "zoomOut", "pdf")
        key(preset, "Plus", "+", shift, "zoomIn", "pdf")
        key(preset, "E", "e", none, "expand", "pdf")
        key(preset, "Tab", "", shift, "focusPrevious", "pdf")
        key(preset, "Tab", "", ctrl, "focusPreview", "preview")
        key(preset, "J", "j", none, "", "pdf")
        key(preset, "K", "k", none, "", "pdf")
        key(preset, "Up", "", none, "cursorUp", "pdf")
        key(preset, "Down", "", none, "cursorDown", "pdf")
        key(preset, "Return", "", none, "open", "pdf")
        key(preset, "Return", "", none, "", "editor")
    }
    key("default", "Return", "", none, "open")
    key("default", "Backspace", "", none, "parent")
    key("default", "H", "H", shift, "historyBack")
    key("default", "L", "L", shift, "historyForward")
    key("default", "Y", "y", none, "copy")
    key("default", "X", "x", none, "cut")
    key("default", "P", "p", none, "paste")
    key("default", "Z", "z", none, "undo")
    key("default", "D", "d", none, "trashArm")
    key("default", "D", "", ctrl, "pageDown")
    key("default", "T", "t", none, "tabNew")
    key("default", "Delete", "", shift, "")
    key("vim", "L", "l", none, "open")
    key("vim", "H", "h", none, "parent")
    key("vim", "Y", "y", none, "copyArm")
    key("vim", "D", "d", none, "cutArm")
    key("vim", "P", "p", none, "pasteArm")
    key("vim", "G", "g", none, "cursorFirstArm")
    key("vim", "G", "G", shift, "cursorLast")
    key("vim", "D", "D", shift, "trash")
    key("vim", "U", "u", none, "undo")
    key("mac", "Return", "", none, "rename")
    key("mac", "Enter", "", none, "rename")
    key("mac", "Right", "", none, "open")
    key("mac", "Left", "", none, "parent")
    key("mac", "C", "", meta, "copy")
    key("mac", "C", "", ctrl, "copy")
    key("mac", "V", "", meta, "paste")
    key("mac", "V", "", ctrl, "paste")
    key("mac", "V", "", meta | alt, "movePaste")
    key("mac", "X", "", ctrl, "")
    key("mac", "D", "", meta, "duplicate")
    key("mac", "Z", "", meta, "undo")
    key("mac", "Z", "", meta | shift, "redo")
    key("mac", "A", "", meta, "selectAll")
    key("mac", "I", "", meta, "properties")
    key("mac", "BracketLeft", "", meta, "historyBack")
    key("mac", "BracketRight", "", meta, "historyForward")
    key("mac", "Period", ".", meta | shift, "toggleHidden")
    key("mac", "Delete", "", shift, "deletePermanently")
    key("mac", "T", "", ctrl, "tabNew")
    key("windows", "Return", "", none, "open")
    key("windows", "F2", "", none, "rename")
    key("windows", "X", "", ctrl, "cut")
    key("windows", "D", "", ctrl, "trash")
    key("windows", "Delete", "", shift, "deletePermanently")
    key("windows", "Y", "", ctrl, "redo")
    key("windows", "Return", "", alt, "properties")
    key("windows", "Enter", "", alt, "properties")
    key("windows", "Left", "", alt, "historyBack")
    key("windows", "Right", "", alt, "historyForward")
    key("windows", "Up", "", alt, "parent")
    key("windows", "H", "", ctrl, "toggleHidden")
    key("windows", "T", "", ctrl, "tabNew")
    Keymap.setPreset("mac")
    check("Mac menu advertises its actual Return action", Keymap.hintFor("rename"), "enter")
    check("Mac Open hint uses native Right", Keymap.hintFor("open"), "right")
    Keymap.setPreset("vim")
    check("Vim Copy hints the key it starts with", Keymap.hintFor("copy"), "y")
    check("Vim Cut hints the key it starts with", Keymap.hintFor("cut"), "d")
    Keymap.setPreset("unknown")
    check("unknown stored preset resolves to Default", Keymap.preset, "default")
    check("Default Copy hint remains y", Keymap.hintFor("copy"), "y")
    // Menus.html and the OpenWith overseer board both draw this row with d; the sheet below keeps dd.
    check("Default Trash hints the key it starts with", Keymap.hintFor("trash"), "d")
    check("Default sheet never advertises a lone destructive d", Keymap.sheetFor("default", "gui").filter(function (row) {
        return row.action === "trash"
    })[0].keys.split(" / ").indexOf("d"), -1)
    check("menu-only actions invent no shortcut", Keymap.hintFor("emptyTrash"), "")
    check("sheet is populated from effective current bindings", Keymap.SHEET.length > 30, true)
    // One cap names one key. Joining every spelling an action answers to produced caps of 40
    // characters on Default and 78 on Mac, wider than the card, and they drew over the next column.
    var widestCap = 0, identifierLabel = ""
    for (var p = 0; p < Keymap.PRESETS.length; p++) {
        var sheet = Keymap.sheetFor(Keymap.PRESETS[p], "gui")
        for (var r = 0; r < sheet.length; r++) {
            if (sheet[r].keys.length > widestCap) widestCap = sheet[r].keys.length
            if (/[a-z][A-Z]/.test(sheet[r].label)) identifierLabel = sheet[r].label
            if (sheet[r].keys.split(" / ").length > 2) identifierLabel = "too many spellings: " + sheet[r].keys
        }
    }
    check("no cap in any preset outgrows its half of the card", widestCap <= 18, true)
    check("no row prints an action id where its wording belongs", identifierLabel, "")
    check("pointer contract remains populated", Keymap.POINTER.length > 10, true)
    var effective = Keymap.bindingRows("mac", "gui")
    check("suppressed Mac Ctrl+X never appears in sheet", effective.some(function (r) { return r.mods === "ctrl" && r.key === "X" }), false)
    check("every effective Mac binding resolves to advertised action", effective.every(function (r) {
        return Keymap.lookupFor("mac", r.keycode, r.text, r.mask, "listing", "gui") === r.action
    }), true)
}

File: src/tui/keymap.rs
use super::input::Key;
use std::collections::HashMap;

#[derive(Default)]
pub struct Map {
    blocks: Vec<(String, HashMap<String, String>)>,
}
impl Map {
    pub fn load() -> Self {
        Self::parse(include_str!("../../keys.toml"))
    }
    // Sample input: [[text]]# comment, followed by char = "j" and action = "cursorDown".
    fn parse(text: &str) -> Self {
        let mut map = Self::default();
        for line in text.lines().map(str::trim) {
            if let Some((kind, suffix)) = line
                .strip_prefix("[[")
                .and_then(|header| header.split_once("]]"))
            {
                if suffix.trim().is_empty() || suffix.trim_start().starts_with('#') {
                    map.blocks.push((kind.trim().into(), HashMap::new()));
                    continue;
                }
            }
            if line.starts_with('#') {
                continue;
            }
            if let Some((key, value)) = line.split_once('=') {
                let value = value.trim();
                if let Ok(crate::jsondoc::Json::Str(value)) = crate::jsondoc::parse(value) {
                    if let Some((_, block)) = map.blocks.last_mut() {
                        block.insert(key.trim().into(), value);
                    }
                }
            }
        }
        map
    }
    pub fn action(&self, key: &Key, preset: &str) -> String {
        self.in_context(key, preset, "listing")
    }
    pub fn in_context(&self, key: &Key, preset: &str, context: &str) -> String {
        if context == "listing" && key.name == "Insert" && key.mods == "ctrl" {
            return "copy".into();
        }
        if context == "listing" && key.name == "Insert" && key.mods == "shift" {
            return "paste".into();
        }
        if preset == "mac" && key.mods == "ctrl" && key.name == "X" {
            return String::new();
        }
        for name in [preset, "all"] {
            for (kind, block) in &self.blocks {
                let here = get(block, "context");
                let mods = get(block, "mods");
                let key_matches = if mods == "text" {
                    matches!(key.mods.as_str(), "" | "shift") && get(block, "key") == key.text
                } else {
                    (if mods == "none" { "" } else { mods }) == key.mods
                        && get(block, "key") == key.name
                };
                if kind == "preset"
                    && get(block, "name") == name
                    && get(block, "frontend") != "gui"
                    && (here.split(',').any(|part| part == context || part == "all")
                        || (here.is_empty() && context == "listing"))
                    && key_matches
                {
                    return get(block, "action").into();
                }
            }
        }
        if context != "listing" {
            return String::new();
        }
        for (kind, block) in &self.blocks {
            let matches = if key.mods.is_empty() {
                (kind == "text" && get(block, "char") == key.text && !key.text.is_empty())
                    || (kind == "code" && get(block, "key") == key.name)
            } else {
                kind == &key.mods && get(block, "key") == key.name
            };
            if matches {
                return get(block, "action").into();
            }
        }
        String::new()
    }
    pub fn sheet(&self, preset: &str) -> Vec<String> {
        let mut rows = Vec::new();
        for (_, label) in self.blocks.iter().filter(|(kind, _)| kind == "sheet") {
            let action = get(label, "action");
            if !supported(action) {
                continue;
            }
            let mut keys = Vec::new();
            for (kind, block) in &self.blocks {
                let mods = if kind == "preset" {
                    get(block, "mods")
                } else {
                    kind
                };
                let name = if kind == "text" {
                    get(block, "char")
                } else {
                    get(block, "key")
                };
                if name.is_empty()
                    || !matches!(
                        mods,
                        "text"
                            | "code"
                            | "none"
                            | ""
                            | "shift"
                            | "ctrl"
                            | "ctrlshift"
                            | "alt"
                            | "super"
                            | "supershift"
                            | "superalt"
                    )
                {
                    continue;
                }
                let text_key = mods == "text";
                let key = if text_key {
                    Key::character(name.chars().next().unwrap(), "")
                } else {
                    Key::named(
                        name,
                        if matches!(mods, "code" | "none") {
                            ""
                        } else {
                            mods
                        },
                    )
                };
                let found = self.action(&key, preset);
                if canonical(&found) != action {
                    continue;
                }
                let prefix = match key.mods.as_str() {
                    "ctrl" => "Ctrl+",
                    "ctrlshift" => "Ctrl+Shift+",
                    "alt" => "Alt+",
                    "super" => "Super+",
                    "supershift" => "Super+Shift+",
                    "superalt" => "Super+Alt+",
                    "shift" => "Shift+",
                    _ => "",
                };
                let chord = format!(
                    "{}{}{}",
                    prefix,
                    name,
                    if found.ends_with("Arm") { name } else { "" }
                );
                if !keys.contains(&chord) {
                    keys.push(chord);
                }
            }
            if !keys.is_empty() {
                rows.push(format!("{}  {}", keys.join(" / "), get(label, "label")));
            }
        }
        rows.push("1–9  Select tab".into());
        rows.push("q  Quit".into());
        rows
    }
}
fn canonical(action: &str) -> &str {
    match action {
        "copyArm" => "copy",
        "cutArm" => "cut",
        "pasteArm" => "paste",
        "cursorFirstArm" => "cursorFirst",
        "trashArm" => "trash",
        _ => action,
    }
}
fn supported(action: &str) -> bool {
    matches!(
        action,
        "cursorDown"
            | "cursorUp"
            | "extendDown"
            | "extendUp"
            | "pageDown"
            | "pageUp"
            | "cursorFirst"
            | "cursorLast"
            | "parent"
            | "historyBack"
            | "historyForward"
            | "open"
            | "toggleSelect"
            | "selectAll"
            | "toggleHidden"
            | "pathBar"
            | "filter"
            | "search"
            | "rename"
            | "newFolder"
            | "newFile"
            | "properties"
            | "copy"
            | "cut"
            | "paste"
            | "movePaste"
            | "duplicate"
            | "trash"
            | "deletePermanently"
            | "undo"
            | "redo"
            | "sortNext"
            | "sortReverse"
            | "tabNew"
            | "openTerminal"
            | "windowNew"
            | "tabClose"
            | "tabNext"
            | "tabPrevious"
            | "preview"
            | "togglePreview"
            | "loadPreview"
            | "focusPreview"
            | "focusNext"
            | "menu"
            | "keymapSheet"
            | "reveal"
            | "escape"
            | "quit"
    )
}
fn get<'a>(map: &'a HashMap<String, String>, key: &str) -> &'a str {
    map.get(key).map(String::as_str).unwrap_or("")
}
#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn compiled_source_drives_text_and_native_ctrl() {
        let map = Map::load();
        assert_eq!(
            map.action(
                &Key {
                    name: "J".into(),
                    text: "j".into(),
                    mods: "".into(),
                    pointer: None,
                },
                "default"
            ),
            "cursorDown"
        );
        assert_eq!(
            map.action(
                &Key {
                    name: "X".into(),
                    text: "X".into(),
                    mods: "ctrl".into(),
                    pointer: None,
                },
                "mac"
            ),
            ""
        );
        assert_eq!(
            map.action(
                &Key {
                    name: "Insert".into(),
                    text: "".into(),
                    mods: "shift".into(),
                    pointer: None,
                },
                "vim"
            ),
            "paste"
        );
    }
    #[test]
    fn contexts_share_only_declared_actions_and_sheet_matches_preset() {
        let map = Map::load();
        assert_eq!(map.in_context(&Key::character('d', ""), "vim", "pdf"), "");
        assert_eq!(
            map.in_context(&Key::named("Space", ""), "mac", "pdf"),
            "preview"
        );
        assert_eq!(
            map.in_context(&Key::named("Tab", "shift"), "windows", "media"),
            "focusPrevious"
        );
        assert_eq!(
            map.in_context(&Key::named("Insert", "ctrl"), "default", "editor"),
            ""
        );
        assert_eq!(map.action(&Key::character('y', ""), "vim"), "copyArm");
        assert!(map.sheet("vim").iter().any(|line| line.contains("yy")));
        assert!(!map
            .sheet("default")
            .iter()
            .any(|line| line.to_lowercase().contains("grid")));
    }
    #[test]
    fn commented_headers_keep_shipped_mac_navigation_aliases_separate() {
        let map = Map::load();
        for (key, action) in [("Up", "parent"), ("Down", "open"), ("Delete", "trash")] {
            assert_eq!(map.action(&Key::named(key, "ctrl"), "mac"), action);
        }
        assert_eq!(map.action(&Key::character('3', "ctrl"), "mac"), "viewGrid");
    }
    #[test]
    fn header_comments_do_not_strip_quoted_hashes_or_accept_other_suffixes() {
        let map = Map::parse(r##"
[[text]] # A header comment can contain = and ]].
char = "#"
action = "literal#action"
# [[preset]] is still a comment.
[[text]]# No intervening space is required.
char = "j"
action = "cursorDown"
[[code]] unexpected
"##);
        assert_eq!(map.blocks.len(), 2);
        assert_eq!(map.action(&Key::character('#', ""), "default"), "literal#action");
        assert_eq!(map.action(&Key::character('j', ""), "default"), "cursorDown");
    }
}
